File: src-tauri/src/models/attempt.rs
use serde::{Deserialize, Serialize};
use specta::Type;
use strum::{Display, EnumString};

#[derive(Serialize, Deserialize, Type, Debug, Clone, PartialEq, Display, EnumString)]
#[serde(rename_all = "PascalCase")]
#[strum(serialize_all = "PascalCase")]
pub enum LiftType {
    Squat,
    Bench,
    Deadlift,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, PartialEq, Display, EnumString)]
#[serde(rename_all = "PascalCase")]
#[strum(serialize_all = "PascalCase")]
pub enum AttemptStatus {
    Pending,
    Successful,
    Failed,
    Skipped,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Attempt {
    pub id: String,
    pub registration_id: String,
    pub lift_type: LiftType,
    pub attempt_number: i32,
    pub weight: f64,
    pub status: AttemptStatus,
    pub timestamp: Option<String>,
    pub judge1_decision: Option<bool>,
    pub judge2_decision: Option<bool>,
    pub judge3_decision: Option<bool>,
    pub notes: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AttemptUpsert {
    pub registration_id: String,
    pub lift_type: LiftType,
    pub attempt_number: i32,
    pub weight: f64,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AttemptUpdateResult {
    pub attempt_id: String,
    pub status: AttemptStatus,
}

File: werewolf-cloud/packages/domain/src/models/attempt.ts
import { z } from 'zod';

export const attemptStatusSchema = z.enum(['Pending', 'Successful', 'Failed']);
export type AttemptStatus = z.infer<typeof attemptStatusSchema>;

export const liftTypeSchema = z.enum(['Squat', 'Bench', 'Deadlift']);
export type LiftType = z.infer<typeof liftTypeSchema>;

export const attemptNumberSchema = z.union([
  z.literal(1),
  z.literal(2),
  z.literal(3),
  z.literal(4),
]);
export type AttemptNumber = z.infer<typeof attemptNumberSchema>;

export const attemptSchema = z.object({
  id: z.string().uuid(),
  registrationId: z.string().uuid(),
  liftType: liftTypeSchema,
  attemptNumber: attemptNumberSchema,
  weight: z.number().nonnegative(),
  status: attemptStatusSchema,
  timestamp: z.string().datetime().nullable().optional(),
  judge1Decision: z.boolean().nullable().optional(),
  judge2Decision: z.boolean().nullable().optional(),
  judge3Decision: z.boolean().nullable().optional(),
  notes: z.string().nullable().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});
export type Attempt = z.infer<typeof attemptSchema>;

export const attemptUpsertSchema = attemptSchema.pick({
  registrationId: true,
  liftType: true,
  attemptNumber: true,
  weight: true,
});
export type AttemptUpsertInput = z.infer<typeof attemptUpsertSchema>;

export const attemptResultUpdateSchema = z.object({
  status: attemptStatusSchema,
  judge1Decision: z.boolean().nullable().optional(),
  judge2Decision: z.boolean().nullable().optional(),
  judge3Decision: z.boolean().nullable().optional(),
  notes: z.string().nullable().optional(),
});
export type AttemptResultUpdate = z.infer<typeof attemptResultUpdateSchema>;
